core/src/main/java/io/questdb/cairo/mv/MatViewRefreshJob.java

```diff
@@ -171,6 +170,107 @@ private static void unionTxnIntervals(LongList txnIntervals, long lo, long hi) {
         }
     }
 
+    private void cacheTxnIntervals(@NotNull TableToken viewToken) {
+        final MatViewState viewState = stateStore.getViewState(viewToken);
+        if (viewState != null && !viewState.isPendingInvalidation() && !viewState.isInvalid() && !viewState.isDropped()) {
+            if (!viewState.tryLock()) {
+                LOG.debug().$("skipping WAL txn intervals caching, locked by a refresh run [view=").$(viewToken).I$();
+                stateStore.enqueueCacheTxnIntervals(viewToken);
+                return;
+            }
+
+            try (WalWriter walWriter = engine.getWalWriter(viewToken)) {
+                final TableToken baseTableToken = verifyBaseTableToken(viewState, walWriter);
+                if (baseTableToken == null) {
+                    return;
+                }
+
+                try (TableReader baseTableReader = engine.getReader(baseTableToken)) {
+                    cacheTxnIntervals0(baseTableReader, viewState, walWriter);
+                } catch (Throwable th) {
+                    LOG.error()
+                            .$("could not cache WAL txn intervals for view [view=").$(viewToken)
+                            .$(", baseTable=").$(baseTableToken)
+                            .$(", ex=").$(th)
+                            .I$();
+                    refreshFailState(viewState, walWriter, th);
+                }
+            } catch (Throwable th) {
+                // If we're here, we couldn't obtain the WAL writer.
+                // Update the in-memory state and call it a day.
+                LOG.error()
+                        .$("could not cache WAL txn intervals, unexpected error [view=").$(viewToken)
+                        .$(", ex=").$(th)
+                        .I$();
+                refreshFailState(viewState, null, th);
+            } finally {
+                viewState.unlock();
+                viewState.tryCloseIfDropped();
+            }
+        }
+    }
+
+    private LongList cacheTxnIntervals0(
+            @NotNull TableReader baseTableReader,
+            @NotNull MatViewState viewState,
+            @NotNull WalWriter walWriter
+    ) {
+        assert viewState.isLocked();
+
+        final long lastTxn = baseTableReader.getSeqTxn();
+        final TableToken baseTableToken = baseTableReader.getTableToken();
+        final MatViewDefinition viewDefinition = viewState.getViewDefinition();
+        final TableToken viewToken = viewDefinition.getMatViewToken();
+        final long lastRefreshTxn = Math.max(viewState.getLastRefreshBaseTxn(), viewState.getCachedIntervalsBaseTxn());
+
+        if (lastRefreshTxn > -1) {
+            if (lastRefreshTxn == lastTxn) {
+                return viewState.getCachedTxnIntervals();
+            }
+
+            try {
```

Comment:
Shall we handle `lastRefreshTxn > lastTxn`? This can happen if the base table is renamed and it is not the same base table as it used to be. We should invalidate the view in this case.